### PR TITLE
feat: add publisher caps to the validator

### PR DESCRIPTION
**Pull request by guibescos**
Computes the `PublisherStakeCapsMessage` no matter what, and adds it to the accumulator Merkle tree if the feature flag is turned on. The idea is that if there are any bugs they should be caught when each individual validator upgrades and not when the flag is turned on.

The computation is performed by https://github.com/pyth-network/pyth-client/pull/412

### CODE REVIEW

File: runtime/src/bank/pyth_accumulator.rs [hunk 1]
```diff
@@ -367,3 +382,32 @@ pub fn update_v2(bank: &Bank) -> std::result::Result<(), AccumulatorUpdateErrorV
...
+    accounts : &Vec<(Pubkey, AccountSharedData)>, 
+    timestamp : i64,
+    messages : &mut Vec<Vec<u8>>,
+    is_active : bool
+) {
+    let parameters: StakeCapParameters = {
+        let data = bank
+        .get_account_with_fixed_root(&STAKE_CAPS_PARAMETERS_ADDR)
+        .unwrap_or_default();
+    let data = data.data();
```
Comment (Riateche) [inline runtime/src/bank/pyth_accumulator.rs]: I think this file is not auto-formatted properly.

File: runtime/src/bank/pyth_accumulator.rs [hunk 2]
```diff
@@ -367,3 +382,32 @@ pub fn update_v2(bank: &Bank) -> std::result::Result<(), AccumulatorUpdateErrorV
...
+    solana_sdk::borsh::try_from_slice_unchecked(data)
+        .unwrap_or(StakeCapParameters { _discriminator: 0, _current_authority : [0u8;32], z: 1, m : 1_000_000_000 })
```
Comment (Riateche) [inline runtime/src/bank/pyth_accumulator.rs]: Maybe implement `Default` for `StakeCapParameters`?

File: runtime/src/bank/pyth_accumulator.rs [hunk 3]
```diff
@@ -367,3 +382,32 @@ pub fn update_v2(bank: &Bank) -> std::result::Result<(), AccumulatorUpdateErrorV
...
+    info!("Computing publisher stake caps with m : {} and z : {}", parameters.m, parameters.z);
+    let account_datas : Vec<&[u8]> = accounts
+        .iter()
+        .map(|(_, account)| account.data().borrow())
+        .collect();
+    let message = pyth_oracle::validator::compute_publisher_stake_caps(account_datas, timestamp, parameters.m, parameters.z);
+
+    if is_active {
+        info!("Adding publisher stake caps to the accumulator");
+        messages.push(message);
```
Comment (Riateche) [inline runtime/src/bank/pyth_accumulator.rs]: I think it would be cleaner to return `Some(message)` from the function instead of passing `&mut Vec` as an argument.

File: runtime/src/bank/pyth_accumulator.rs [hunk 4]
```diff
@@ -367,3 +382,32 @@ pub fn update_v2(bank: &Bank) -> std::result::Result<(), AccumulatorUpdateErrorV
...
+    solana_sdk::borsh::try_from_slice_unchecked(data)
+        .unwrap_or(StakeCapParameters { _discriminator: 0, _current_authority : [0u8;32], z: 1, m : 1_000_000_000 })
+};
+
+    info!("Computing publisher stake caps with m : {} and z : {}", parameters.m, parameters.z);
+    let account_datas : Vec<&[u8]> = accounts
+        .iter()
+        .map(|(_, account)| account.data().borrow())
+        .collect();
+    let message = pyth_oracle::validator::compute_publisher_stake_caps(account_datas, timestamp, parameters.m, parameters.z);
```
Comment (Riateche) [inline runtime/src/bank/pyth_accumulator.rs]: Is there a point in computing this if `is_active` is false?

File: runtime/src/bank/pyth_accumulator.rs [hunk 5]
```diff
@@ -367,3 +382,32 @@ pub fn update_v2(bank: &Bank) -> std::result::Result<(), AccumulatorUpdateErrorV
 
     update_v1(bank, &v2_messages, any_v1_aggregations)
 }
+
+pub fn compute_publisher_stake_caps(
+    bank : &Bank,
+    accounts : &Vec<(Pubkey, AccountSharedData)>,
```
Comment (Riateche) [inline runtime/src/bank/pyth_accumulator.rs]: Usually `&[T]` is preferred over `&Vec<T>`.

File: runtime/src/bank/pyth_accumulator.rs [hunk 6]
```diff
@@ -341,6 +353,9 @@ pub fn update_v2(bank: &Bank) -> std::result::Result<(), AccumulatorUpdateErrorV
     let mut any_v1_aggregations = false;
     let mut v2_messages = Vec::new();
 
+    let add_publisher_stake_caps = bank.feature_set.is_active(&feature_set::add_publisher_stake_caps_to_the_accumulator::id());
```
Comment (Riateche) [inline runtime/src/bank/pyth_accumulator.rs]: I think moving the feature flag check inside `compute_publisher_stake_caps()` might make sense.

File: runtime/src/bank/pyth_accumulator.rs [hunk 7]
```diff
@@ -367,3 +382,32 @@ pub fn update_v2(bank: &Bank) -> std::result::Result<(), AccumulatorUpdateErrorV
 
     update_v1(bank, &v2_messages, any_v1_aggregations)
 }
+
+pub fn compute_publisher_stake_caps(
```
Comment (ali-behjati) [inline runtime/src/bank/pyth_accumulator.rs:466]: it's cleaner if you call this one if the feature is active and it returns a Message instead of mutating the vector.

File: runtime/src/bank/pyth_accumulator.rs [hunk 8]
```diff
@@ -367,3 +382,32 @@ pub fn update_v2(bank: &Bank) -> std::result::Result<(), AccumulatorUpdateErrorV
...
+    let parameters: StakeCapParameters = {
+        let data = bank
+        .get_account_with_fixed_root(&STAKE_CAPS_PARAMETERS_ADDR)
+        .unwrap_or_default();
+    let data = data.data();
+    solana_sdk::borsh::try_from_slice_unchecked(data)
+        .unwrap_or(StakeCapParameters { _discriminator: 0, _current_authority : [0u8;32], z: 1, m : 1_000_000_000 })
+};
+
+    info!("Computing publisher stake caps with m : {} and z : {}", parameters.m, parameters.z);
```
Comment (ali-behjati) [inline runtime/src/bank/pyth_accumulator.rs]: I recommend prefixing your logs with something so we can separate them easier.

**Review by Riateche — CHANGES_REQUESTED** [review]

File: runtime/src/bank/pyth_accumulator.rs (hunk 8 above)
Comment (guibescos) [reply runtime/src/bank/pyth_accumulator.rs]: done

File: runtime/src/bank/pyth_accumulator.rs (hunk 3 above)
Comment (guibescos) [reply runtime/src/bank/pyth_accumulator.rs]: done

File: runtime/src/bank/pyth_accumulator.rs (hunk 5 above)
Comment (guibescos) [reply runtime/src/bank/pyth_accumulator.rs]: fixed

File: runtime/src/bank/pyth_accumulator.rs (hunk 6 above)
Comment (guibescos) [reply runtime/src/bank/pyth_accumulator.rs]: fixed

File: runtime/src/bank/pyth_accumulator.rs (hunk 4 above)
Comment (guibescos) [reply runtime/src/bank/pyth_accumulator.rs]: I want to compute this if `is_active` is false so that if there's a bug or it crashes during the computation, we realize when validators update individually and not when the flag gets activated. But happy to change this.

File: runtime/src/bank/pyth_accumulator.rs (hunk 7 above)
Comment (guibescos) [reply runtime/src/bank/pyth_accumulator.rs:466]: fixed

File: runtime/src/bank/pyth_accumulator.rs (hunk 2 above)
Comment (guibescos) [reply runtime/src/bank/pyth_accumulator.rs]: Done

File: runtime/src/bank/pyth_accumulator.rs (hunk 1 above)
Comment (guibescos) [reply runtime/src/bank/pyth_accumulator.rs]: Done

File: runtime/src/bank/pyth_accumulator.rs [hunk 9]
```diff
@@ -366,3 +382,36 @@ pub fn update_v2(bank: &Bank) -> std::result::Result<(), AccumulatorUpdateErrorV
...
+        parameters.m, parameters.z
+    );
+    let message = pyth_oracle::validator::compute_publisher_stake_caps(
+        accounts.iter().map(|(_, account)| account.data()),
+        bank.clock().unix_timestamp,
+        parameters.m,
+        parameters.z,
+    );
+
+    if bank
```
Comment (OpenCoreCH) [inline runtime/src/bank/pyth_accumulator.rs:495]: Small detail: Checking this at the beginning and exiting early if the feature set is deactivated would avoid some unnecessary computations if it is deactivated.

File: validator/src/main.rs [hunk 10]
```diff
@@ -3177,9 +3177,11 @@ fn process_account_indexes(matches: &ArgMatches) -> AccountSecondaryIndexes {
         })
         .collect();
 
-    assert!(account_indexes.contains(&AccountIndex::ProgramId),
+    assert!(
```
Comment (guibescos) [inline validator/src/main.rs:3180]: drive-by format

File: runtime/src/bank/pyth_accumulator.rs (hunk 9 above)
Comment (guibescos) [reply runtime/src/bank/pyth_accumulator.rs:495]: It can be useful so that we detect errors in the logic before the feature activation or before a majority of nodes have updated.

**Review by keyvankhademi — APPROVED** [review]

**Review by Riateche — APPROVED** [review]

File: runtime/Cargo.toml [hunk 11]
```diff
@@ -35,8 +35,8 @@ num-traits = { version = "0.2" }
 num_cpus = "1.13.1"
 once_cell = "1.12.0"
 ouroboros = "0.15.0"
-pyth-oracle = { git = "https://github.com/pyth-network/pyth-client", tag = "oracle-v2.32.1", features = ["library"] }
-pythnet-sdk = { git = "https://github.com/pyth-network/pyth-crosschain", version = "1.13.6", rev = "e670f57f89b05398ca352e4accb1e32724a8e1b4" }
+pyth-oracle = { git = "https://github.com/pyth-network/pyth-client.git" ,branch="feat/publisher-caps" , features = ["library"] }
```
Comment (ali-behjati) [inline runtime/Cargo.toml]: let's change it once it's merged.

**Review by ali-behjati — APPROVED** [review]
